confidant/public/modules/history/services/archive.js

```diff
@@ -10,53 +10,148 @@
+                    resource.type = type;
+                    resource.modified_date = new Date(resource.modified_date);
+                    if (type === 'services') {
+                        // For consistency, everything has a name, so we'll set the service name from its ID.
+                        // The ID is name-revision, so we need to split and take the first item.
+                        resource.name = resource.id.split('-')[0];
                     }
-                    for (i = blindCredentialArchive.length; i--;) {
-                        blindCredentialArchive[i].type = 'blind_credential';
+                });
+            }
+
+            function queryUntilItem(service, type, item, data, page, count) {
+                data = data || [];
+                page = page || null;
+                count = count || 0;
+                var max = 10,
+                    defer = $q.defer();
+
+                doQuery(service, page).then(function(result) {
+                    var matchIndex = null,
+                        newData = result[type];
+                    for (var i = newData.length; i--;) {
+                        if (newData[i].id === item.id) {
+                            matchIndex = i;
+                            break;
+                        }
                     }
-                    for (i = serviceArchive.length; i--;) {
-                        serviceArchive[i].type = 'service';
-                        var nameArr = serviceArchive[i].id.split('-');
-                        nameArr.pop();
-                        serviceArchive[i].name = nameArr.join('-');
+                    if (matchIndex === null) {
+                        if (count > max) {
+                            // Sanity check base case. If we have an issue with eventual consistency
+                            // on the backend, let's not recurse forever.
+                            defer.resolve([]);
+                        } else {
+                            newData = data.concat(newData);
+                            queryUntilItem(service, type, item, newData, result.next_page, count+1).then(function (chainedResult) {
+                                defer.resolve(chainedResult);
+                            });
+                        }
+                    } else {
+                        // base case
+                        defer.resolve(newData.slice(0, matchIndex));
                     }
-                    _this.resourceArchive = credentialArchive.concat(blindCredentialArchive);
-                    _this.resourceArchive = _this.resourceArchive.concat(serviceArchive);
-                    _this.resourceArchive.forEach(function(resource){
-                        resource.modified_date = new Date(resource.modified_date);
-                    });
+                });
+
+                return defer.promise;
+            }
+
+            this.setLimit = function(limit) {
+                _this.limit = limit;
```

Comment:
curious, what does this do?  wondering why we need a wrapper to do this.

Reply:
You can't access the scope from within a service, but we get the limit from the server side, which gets added to the scope at the top level app config: https://github.com/lyft/confidant/blob/master/confidant/public/modules/app.js#L24-L34